File: linda-core/src/main/java/it/unibo/coordination/linda/core/NegatedTupleSpace.java
package it.unibo.coordination.linda.core;

import java.util.Optional;
import java.util.concurrent.CompletableFuture;

public interface NegatedTupleSpace<T extends Tuple, TT extends Template, K, V> extends TupleSpace<T, TT, K, V> {
    CompletableFuture<Match<T, TT, K, V>> absent(TT template);

    default CompletableFuture<TT> absentTemplate(TT template) {
        return absent(template).thenApplyAsync(Match::getTemplate);
    }

    CompletableFuture<Match<T, TT, K, V>> tryAbsent(TT template);

    default CompletableFuture<Optional<T>> tryAbsentTuple(TT template) {
        return tryAbsent(template).thenApplyAsync(Match::getTuple);
    }
}

File: linda-core/src/main/kotlin/it/unibo/coordination/linda/core/NegatedTupleSpace.kt
package it.unibo.coordination.linda.core

import it.unibo.coordination.Promise
import java.util.*

interface NegatedTupleSpace<T : Tuple<T>, TT : Template<T>, K, V, M : Match<T, TT, K, V>> : LindaTupleSpace<T, TT, K, V, M> {

    fun absent(template: TT): Promise<M>

    @JvmDefault
    fun absent(template: String): Promise<M> =
            absent(template.toTemplate())

    @JvmDefault
    fun absentTemplate(template: TT): Promise<TT> {
        return absent(template).thenApplyAsync { it.template }
    }

    @JvmDefault
    fun absentTemplate(template: String): Promise<TT> =
            absentTemplate(template.toTemplate())

    fun tryAbsent(template: TT): Promise<M>

    @JvmDefault
    fun tryAbsent(template: String): Promise<M> =
            tryAbsent(template.toTemplate())

    @JvmDefault
    fun tryAbsentTuple(template: TT): Promise<Optional<T>> {
        return tryAbsent(template).thenApplyAsync { it.tuple }
    }

    @JvmDefault
    fun tryAbsentTuple(template: String): Promise<Optional<T>> =
            tryAbsentTuple(template.toTemplate())
}